### PR TITLE
Fast differentiable sorting and ranking

### PR DESCRIPTION
**Pull request by BatyLeo**
Following this [discourse post](https://discourse.julialang.org/t/do-we-have-differentiable-sort-in-julia/102713/1), here is a Julia implementation of [Fast differentiable sorting and ranking](https://arxiv.org/pdf/2002.08871.pdf) (see [here](https://github.com/google-research/fast-soft-sort/tree/master) for the original Python implementation).

This includes:
- `soft_rank` and `soft_sort` differentiable methods
- options for choosing regularization: l2 or entropic
- Two new custom optimization layers:`SoftRank <: AbstractRegularized` and `SoftSort <: AbstractOptimizationLayer` compatible with the InferOpt ecosystem

---

Reproducing figures from the paper:

```julia
using InferOpt, Plots, Zygote
```

```julia
plot(x -> sort([0.0, x, 1.0, 2.0]; rev=true)[2], label="Sort")
plot!(x -> soft_sort_l2([0.0, x, 1.0, 2.0]; rev=true)[2], label="Soft sort l2")
plot!(y -> gradient(x -> soft_sort_l2([0.0, x, 1.0, 2.0]; ε=1.0, rev=true)[2], y)[1], label="Soft sort l2 derivative")
```
plot

```julia
plot(x -> sort([0.0, x, 1.0, 2.0]; rev=true)[2], label="Sort")
plot!(x -> soft_sort_kl([0.0, x, 1.0, 2.0]; rev=true)[2], label="Soft sort kl")
plot!(y -> gradient(x -> soft_sort_kl([0.0, x, 1.0, 2.0]; ε=1.0, rev=true)[2], y)[1], label="Soft sort kl derivative")
```
plot

```julia
plot(x -> ranking([x, 3.0, 1.0, 2.0]; rev=true)[1], label="Ranking")
plot!(x -> soft_rank_l2([x, 3.0, 1.0, 2.0]; rev=true)[1], label="Soft rank l2")
plot!(y -> gradient(x -> soft_rank_l2([x, 3.0, 1.0, 2.0]; ε=1.0, rev=true)[1], y)[1], label="Soft rank l2 derivative")
```
plot

```julia
plot(x -> ranking([x, 3.0, 1.0, 2.0]; rev=true)[1], label="Ranking")
plot!(x -> soft_rank_kl([x, 3.0, 1.0, 2.0]; rev=true)[1], label="Soft rank kl")
plot!(y -> gradient(x -> soft_rank_kl([x, 3.0, 1.0, 2.0]; ε=1.0, rev=true)[1], y)[1], label="Soft rank kl derivative")
```
plot

### CODE REVIEW

File: src/regularized/soft_rank.jl [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+- `rev::Bool=false`: rank in ascending order if false
+- `is_l2_regularized::Bool=true`: use l2 regularization if true, else kl regularization
+"""
+function SoftRank(; ε::Float64=1.0, rev::Bool=false, is_l2_regularized::Bool=true)
+    return SoftRank{is_l2_regularized}(ε, rev)
+end
+
+(l::SoftRank{true})(θ; ε=l.ε, rev=l.rev) = soft_rank_l2(θ; ε, rev)
+(l::SoftRank{false})(θ; ε=l.ε, rev=l.rev) = soft_rank_kl(θ; ε, rev)
+compute_regularization(l::SoftRank, y) = l.ε * half_square_norm(y)
```
Comment (LouisBouvier) [inline src/regularized/soft_rank.jl:37]: The version with KL seems to be missing.

File: test/soft_rank.jl [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        rank_jac = Zygote.jacobian(x -> soft_rank_kl(x; ε=10.0), θ)[1]
+        rank_jac_fd = FiniteDifferences.jacobian(
+            central_fdm(2, 1), x -> soft_rank_kl(x; ε=10.0), θ
+        )[1]
+        @test all(isapprox.(rank_jac, rank_jac_fd, atol=1e-4))
+    end
+end
+
+@testitem "Learn by experience soft rank l2" default_imports = false begin
```
Comment (LouisBouvier) [inline test/soft_rank.jl:60]: What would you say about a comparison with classic Perturbed in terms of computation time and solution quality ?

File: test/soft_rank.jl [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        true_maximizer=ranking,
+        maximizer=soft_rank_kl,
+        loss=cost,
+        error_function=hamming_distance,
+        true_encoder=true_encoder,
+        cost=cost,
+    )
+end
+
+@testitem "Fenchel-Young loss soft rank L2" default_imports = false begin
```
Comment (LouisBouvier) [inline test/soft_rank.jl:98]: Same question as above, can we check performance gets better as we dive into the details of the CO problem to differentiate ?

File: test/soft_rank.jl (hunk 2 above)
Comment (BatyLeo) [reply test/soft_rank.jl:60]: Good idea, I'll look into it!

**Pull request merged**